### PR TITLE
Jkmarx/group patch perms

### PR DESCRIPTION
**Pull request by jkmarx**
Ref #2618
- Extend group api to return all user groups query with data set (all_perms flag)
- Extend group api to patch a group's perms for a data set
- Refactor a utils method to retrieve and raise errors when grabbing data set.
- Unit tests

### CODE REVIEW

File: refinery/core/test_views.py [hunk 1]
```diff
@@ -925,11 +926,161 @@ def test_get_groups_with_data_set_uuid_has_correct_perms_field(self):
...
+        self.assertEqual(patch_response.data.get('perm_list').get('change'),
+                         True)
+        self.assertEqual(patch_response.data.get('perm_list').get('read'),
+                         True)
+        self.assertEqual(patch_response.data.get('perm_list').get('read_meta'),
+                         True)
+
+    def test_patch_groups_only_allows_owners_and_returns_401(self):
+        self.non_owner = User.objects.create_user('Non-owner',
+                                                  'user@fake.com',
```
Comment (hackdna) [inline refinery/core/test_views.py]: 'user@example.com' or similar: https://tools.ietf.org/html/rfc2606#section-3

File: refinery/core/test_views.py [hunk 2]
```diff
@@ -925,11 +926,161 @@ def test_get_groups_with_data_set_uuid_has_correct_perms_field(self):
...
+        self.data_set.unshare(self.group_2)
+        get_request = self.factory.get(self.url_root,
+                                       {'dataSetUuid': self.data_set.uuid,
+                                        'allPerms': True})
+        get_request.user = self.user
+        get_response = self.view(get_request)
+
+        for group in get_response.data:
+            response_perms = group.get('perm_list')
+            if (group.get('id') == self.group.id):
```
Comment (hackdna) [inline refinery/core/test_views.py]: Unnecessary parens

File: refinery/core/test_utils.py [hunk 3]
```diff
@@ -69,3 +72,21 @@ def test_only_non_manager_groups_return_correct_number(self):
...
         for group in get_non_manager_groups_for_user(self.user):
             self.assertFalse(group.extendedgroup.is_manager_group())
+
+
+class GetDataSetForViewSetTest(TestCase):
+    def setUp(self):
+        self.username = 'coffee_lover'
+        self.password = 'coffeecoffee'
+        self.user = User.objects.create_user(self.username,
+                                             'user@fake.com',
```
[the remark this answers is not in the document]
Comment (jkmarx) [reply refinery/core/test_utils.py]: Worth noting we use this example in many places, but sure I'll update it.

File: refinery/core/views.py [hunk 4]
```diff
@@ -925,36 +925,67 @@ def delete(self, request, uuid):
...
         if not ('read_meta_dataset' in get_perms(public_group, data_set) or
                 request.user.has_perm('core.read_meta_dataset', data_set)):
             return Response(data_set_uuid, status=status.HTTP_401_UNAUTHORIZED)
 
-        groups_with_perms = get_groups_with_perms(data_set)
+        if all_perms_flag:
+            member_groups = [group for group in ExtendedGroup.objects.all()
+                             if request.user in group.user_set.all() and
+                             not group.is_manager_group()]
+            serializer = ExtendedGroupSerializer(
```
[the remark this answers is not in the document]
Comment (jkmarx) [reply refinery/core/views.py]: I think for clarity it's fine.

File: refinery/core/views.py [hunk 5]
```diff
@@ -925,36 +925,67 @@ def delete(self, request, uuid):
...
+    def list(self, request):
+        data_set_uuid = request.query_params.get('dataSetUuid')
+        all_perms_flag = request.query_params.get('allPerms', False)
+
+        data_set = get_data_set_for_view_set(data_set_uuid)
 
         public_group = ExtendedGroup.objects.public_group()
         if not ('read_meta_dataset' in get_perms(public_group, data_set) or
                 request.user.has_perm('core.read_meta_dataset', data_set)):
             return Response(data_set_uuid, status=status.HTTP_401_UNAUTHORIZED)
```
[the remark this answers is not in the document]
Comment (jkmarx) [reply refinery/core/views.py]: I'll return a 403.

File: refinery/core/views.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hackdna) [reply refinery/core/views.py]: This line is essentially repeated twice - not great for clarity or maintainability.

File: refinery/core/views.py [hunk 6]
```diff
@@ -925,37 +908,73 @@ def delete(self, request, uuid):
...
+        data_set_uuid = request.query_params.get('dataSetUuid')
+        all_perms_flag = request.query_params.get('allPerms', False)
+
+        data_set = get_data_set_for_view_set(data_set_uuid)
 
         public_group = ExtendedGroup.objects.public_group()
         if not ('read_meta_dataset' in get_perms(public_group, data_set) or
                 request.user.has_perm('core.read_meta_dataset', data_set)):
-            return Response(data_set_uuid, status=status.HTTP_401_UNAUTHORIZED)
+            return Response(data_set_uuid, status=status.HTTP_403_FORBIDDEN)
```
Comment (hackdna) [inline refinery/core/views.py]: Why not raise PermissionDenied?
https://www.django-rest-framework.org/api-guide/exceptions/#permissiondenied

File: refinery/core/views.py [hunk 7]
```diff
@@ -925,37 +908,73 @@ def delete(self, request, uuid):
...
+                                             context={'data_set': data_set})
         return Response(serializer.data)
 
+    def partial_update(self, request, uuid, format=None):
+        data_set_uuid = request.data.get('dataSetUuid')
+        group = self.get_object(uuid)
+
+        data_set = get_data_set_for_view_set(data_set_uuid)
+        if data_set.get_owner() != request.user:
+            return Response(data_set_uuid, status=status.HTTP_403_FORBIDDEN)
```
Comment (hackdna) [inline refinery/core/views.py]: Why not raise PermissionDenied?
https://www.django-rest-framework.org/api-guide/exceptions/#permissiondenied

File: refinery/core/views.py [hunk 8]
```diff
@@ -925,37 +908,73 @@ def delete(self, request, uuid):
...
+            data_set.share(group, False, False)
+        elif group_perm_update.get('read'):
+            data_set.share(group, True, False)
+        elif group_perm_update.get('read_meta'):
+            data_set.share(group, False, True)
+
+        serializer = ExtendedGroupSerializer(group,
+                                             context={'data_set': data_set})
+
+        return Response(serializer.data, status=status.HTTP_200_OK)
```
Comment (hackdna) [inline refinery/core/views.py]: Is it necessary to set HTTP 200 on a response explicitly? It looks like that's the default value: https://www.django-rest-framework.org/api-guide/responses/#response

File: refinery/core/utils.py [hunk 9]
```diff
@@ -1031,3 +1016,14 @@ def _wrapped_view(request, *args, **kwargs):
...
+
+def get_data_set_for_view_set(uuid):
+    try:
+        return core.models.DataSet.objects.get(uuid=uuid)
+    except core.models.DataSet.DoesNotExist as e:
+        logger.error(e)
+        raise Http404
+    except core.models.DataSet.MultipleObjectsReturned as e:
+        logger.error(e)
+        raise APIException("Multiple dataSets returned for this request.")
```
Comment (hackdna) [inline refinery/core/utils.py]: Might be worth simplifying this method using `get_object_or_404()` shortcut: https://stackoverflow.com/questions/22552306/what-will-happen-if-get-object-or-404-finds-many-objects (also, this pattern might be applicable in other similar cases)

**Pull request merged**